### PR TITLE
add before emit hook

### PR DESCRIPTION
**Pull request by barak007**
allow consumers to add additional mapping of packages and create backward compatibility of metadata

### CODE REVIEW

File: packages/webpack-extensions/test/e2e/metadata-plugin-project.spec.ts [hunk 1]
```diff
@@ -1,6 +1,7 @@
 import { StylableProjectRunner } from '@stylable/e2e-test-kit';
 import { expect } from 'chai';
 import { join } from 'path';
+import { ComponentsMetadata } from '@stylable/webpack-extensions/src';
```
Comment (AviVahl) [inline packages/webpack-extensions/test/e2e/metadata-plugin-project.spec.ts:4]: ../../src

File: packages/webpack-extensions/test/e2e/projects/metadata-plugin-project/webpack-before-emit.config.ts [hunk 2]
```diff
@@ -0,0 +1,30 @@
...
+        new StylableWebpackPlugin({
+            optimize: {
+                shortNamespaces: true,
+            },
+        }),
+        new StylableMetadataPlugin({
+            name: 'test',
+            version: '1.0.0',
+            beforeEmit(builder) {
+                const metadata = builder.build();
```
Comment (AviVahl) [inline packages/webpack-extensions/test/e2e/projects/metadata-plugin-project/webpack-before-emit.config.ts:20]: it's odd that a `beforeEmit` hook is the place when `build()` is called.

should it get the `metadata` from parameters? (without full access to the builder)

File: packages/webpack-extensions/test/e2e/projects/metadata-plugin-project/webpack-before-emit.config.ts [hunk 3]
```diff
@@ -0,0 +1,30 @@
...
+const HtmlWebpackPlugin = require('html-webpack-plugin');
+
+module.exports = {
+    mode: 'development',
+    context: __dirname,
+    devtool: 'source-map',
+    plugins: [
+        new StylableWebpackPlugin({
+            optimize: {
+                shortNamespaces: true,
```
Comment (AviVahl) [inline packages/webpack-extensions/test/e2e/projects/metadata-plugin-project/webpack-before-emit.config.ts:13]: is this relevant for the test?

File: packages/webpack-extensions/test/e2e/projects/metadata-plugin-project/webpack-before-emit.config.ts [hunk 4]
```diff
@@ -0,0 +1,30 @@
...
+            },
+        }),
+        new StylableMetadataPlugin({
+            name: 'test',
+            version: '1.0.0',
+            beforeEmit(builder) {
+                const metadata = builder.build();
+                metadata.packages['test'] = '/test';
```
Comment (AviVahl) [inline packages/webpack-extensions/test/e2e/projects/metadata-plugin-project/webpack-before-emit.config.ts:21]: what is this line doing?

File: packages/webpack-extensions/test/e2e/metadata-plugin-project.spec.ts [hunk 5]
```diff
@@ -119,6 +120,27 @@ describe(`(${project})`, () => {
...
+                },
+                configName: 'webpack-before-emit.config',
+            },
+            before,
+            afterEach,
+            after
+        );
+
+        it('contains metadata file with override from beforeEmit', () => {
+            const s = projectRunnerJs.getBuildAsset('test.metadata.json');
```
Comment (AviVahl) [inline packages/webpack-extensions/test/e2e/metadata-plugin-project.spec.ts:138]: s?

**Comment by barak007** [comment]
Will not be used.

**Pull request closed without merging**